### ISSUE TITLE
The mystery of the missing GPM toggle button

### ISSUE THREAD

**Issue by stevehadd** [informatics-lab/forest#59]
The toggle button in GPM/model app to switch between early and late datasets has gone missing in mysterious circumstances. Your mission, should you choose to accept it, it to track the missing button and return to its place in the GPM/model app.

**Comment by andrewgryan** [comment]
The default behaviour of `bokeh.models.RadioButtonGroup` given an empty list is to not add any elements to the HTML document. This seems to be the smoking gun of the GPM mystery

The controller `ModelGpmControl` performs a list comprehension on datasets.keys() to detect keys containing 'imerg' strings. However, the datasets.keys() was changed in `main.py` to have forecast times as keys not early/late runs. Thus an empty list is returned.

To fix this the list comprehension should be modified to detect the correct 'imerg' strings

**Comment by andrewgryan** [comment]
I think I have cured the issue, essentially I wrote unit tests to discover the `datasets` dictionary shape and how things are passed through the various objects in `main.py` and `model_gpm_control.py`

I added this knowledge into the docstrings and developed a parsing method `ModelGpmControl.imerg_labels()` to keep all GPM IMERG Early/Late labels consistent

My developments are in the branch [forest_59_gpm_mystery](/informatics-lab/forest/tree/forest_59_gpm_mystery), let me know if you need further explanation

I don't have data to run the affected app, could you give it a spin and let me know if the GPM button has reappeared?

**Comment by andrewgryan** [comment]
I reckon I have overlooked what is happening on [line 269](https://github.com/informatics-lab/forest/blob/d512c3ebff2fc6d5db012de69f3bdb63f4332a5a/bokeh_apps/plot_sea_model_and_gpm_mpl/model_gpm_control.py#L269)

**Comment by andrewgryan** [comment]
I assumed setting the labels to any "human readable" value was enough. Turns out they are meant to be keys to the datasets dictionaries

The `on_imerg_change` call back calls through to `forest.plot` with the label information and updates the "config" on [forest/plot.py line 102](https://github.com/informatics-lab/forest/blob/d512c3ebff2fc6d5db012de69f3bdb63f4332a5a/forest_lib/forest/plot.py#L102)

**Comment by andrewgryan** [comment]
I think the buttons must have been named with the dictionary keys, e.g. `gpm_imerg_early` and `gpm_imerg_late`. I can revert the code to this behaviour or extend it to use human-readable buttons, e.g. 'GPM IMERG Early', that feed the keywords through the call back

**Comment by andrewgryan** [comment]
Actually, the button label is used to access `data_type_name` to use as a plot description. But I am using `data_type_name` for the button labels so I could replace `self.dataset[self.current_config]['data_type_name']` with the label value

**Comment by andrewgryan** [comment]
I think I have finally solved the problem. The on change call back should now send the appropriate data to `ForestPlot.set_config` while displaying human readable buttons

I used a set of tuples, since ordering was important, to find/keep track of the button data. I introduced some unit tests to aid my investigation work and to provide a guarantee that the GPM buttons won't go astray again in the same way 

Before I finish the branch, I may need to cache the result as accessing `ModelGpmControl.imerg_tuples` traverses all "datasets" every time. Which could be slow if there are many dates available, but a non cached approach keeps the buttons in sync with the available datasets
